toolchain/check/handle_index.cpp

```diff
@@ -26,6 +26,68 @@ auto HandleParseNode(Context& /*context*/, Parse::IndexExprStartId /*node_id*/)
   return true;
 }
 
+// Performs bounds checking for string indexing when the index is a constant.
+static auto CheckStringIndexBounds(Context& context,
+                                   SemIR::InstId operand_inst_id,
+                                   SemIR::InstId index_inst_id,
+                                   const llvm::APInt& index_int) -> void {
+  if (index_int.isNegative()) {
+    CARBON_DIAGNOSTIC(ArrayIndexNegative, Error, "index `{0}` is negative.",
```

Comment:
```suggestion
    CARBON_DIAGNOSTIC(StringIndexNegative, Error, "index `{0}` is negative.",
```
We can't reuse the same diagnostic in more than one place.